### ISSUE TITLE
per sample gradient in torch.autograd.grad

### ISSUE THREAD

**Issue by twangnh** [pytorch/pytorch#132378]
### 🚀 The feature, motivation and pitch

I'm implementing an algorithm that requires per-sample gradient computation w.r.t a network layer weights, currently I implement this in a naive loop fashion, i.e. call `torch.autograd.grad ` for each sample loss, the following is a example code:

```
def compute_grad(loss_batch, weight):
"""
loss_batch, batched loss for multiple samples, [N,]
weight, the network layer weight for computing the gradients
"""
    grads = []
    for i in range(len(loss_bbox)):
        grad_persample = torch.autograd.grad(loss_batch[i], weight, retain_graph=True)[0]
        grads.append(grad_persample)
    grads = torch.stack(grads)
    return grads
```

but this implementation is very slow, so I would like to implement it in batch, but currently `torch.autograd.grad` seems only support scalar input as loss, there are some suggestions from the pytorch Forum [batch per sample gradient](https://discuss.pytorch.org/t/simple-use-case-compete-gradient-independently-for-multiple-sample-loss/207317) that uses torch.func to parallelize the grdient computation, but that turns out cumbersome to implement, loses the simplitiy of calling `torch.autograd.grad` with loss and weight as input.
 so I propose to add a new feature of batched gradient computation with `torch.autograd.grad`

### Alternatives

the torch.func to parallieze the computation, which lost the simplicity of calling `torch.autograd.grad ` and is cumbersome to implement, as mentioned in the [batch per sample gradient](https://discuss.pytorch.org/t/simple-use-case-compete-gradient-independently-for-multiple-sample-loss/207317)

**Comment by soulitzer** [comment]
`torch.autograd.grad` should already support per-sample gradients via the parameter `is_grads_batched=True`, however unlike the recommended torch.func approach, this API uses an older, experimental version of vmap underneath which is no longer actively developed.

**Comment by zou3519** [comment]
You need torch.func (composing torch.func.grad and torch.func.vmap) to compute per-sample gradients. The quantity that torch.autograd.grad computes is "batched gradients", which is not what we want here.

**Comment by twangnh** [comment]
@zou3519 Thanks for the suggestion and I have found this documentation about per sample gradient with vmap and func https://pytorch.org/functorch/stable/notebooks/per_sample_grads.html, but we want to implement the per sample gradient w.r.t a single layer weight, which I seems cumbersome to implement, since most of our torch models are not simple multi-layer networks.
